File: Data Structures & Algorithms/meeting-schedule-ii/submission-0.cpp
/**
 * Definition of Interval:
 * class Interval {
 * public:
 *     int start, end;
 *     Interval(int start, int end) {
 *         this->start = start;
 *         this->end = end;
 *     }
 * }
 */

class Solution {
public:
    int minMeetingRooms(vector<Interval>& intervals) {
        vector<int> starts;
        vector<int> ends;

        for(const auto& i: intervals){
            starts.emplace_back(i.start);
            ends.emplace_back(i.end);

        }
        sort(starts.begin(), starts.end());
        sort(ends.begin(), ends.end());

        int s=0,e=0;
        int ans=0, cnt=0;

        while(s<starts.size()){
            if(starts[s]<ends[e]){
                ++cnt;
                ans=max(ans,cnt);
                s++;
            }
            else{
                --cnt;
                e++;
            }
        }
        return ans;

    }
};

File: Data Structures & Algorithms/merge-intervals/submission-0.java
class Solution {
    public int[][] merge(int[][] intervals) {

        // Sort intervals based on starting value
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        List<int[]> ans = new ArrayList<>();

        // Start with first interval
        int[] prev = intervals[0];

        for (int i = 1; i < intervals.length; i++) {

            int[] curr = intervals[i];

            // Overlapping intervals
            if (curr[0] <= prev[1]) {
                prev[1] = Math.max(prev[1], curr[1]);
            } 
            
            // Non-overlapping interval
            else {
                ans.add(prev);
                prev = curr;
            }
        }

        // Add the last remaining interval
        ans.add(prev);

        return ans.toArray(new int[ans.size()][]);
    }
}